File: Sep/Day19/design-spreadsheet.cc
#include <bits/stdc++.h>
using namespace std;

class Spreadsheet {
private:
    unordered_map<string, int> cellMap;

    vector<string> extractTokens(const string &formula) {
        vector<string> tokens;
        regex re("([A-Z]+\\d+|\\d+)");
        auto begin = sregex_iterator(formula.begin(), formula.end(), re);
        auto end = sregex_iterator();
        for (auto it = begin; it != end; ++it) {
            tokens.push_back(it->str());
        }
        return tokens;
    }

public:
    // Version 1: pre-populate cellMap
    Spreadsheet(int rows, bool prepopulate) {
        if (prepopulate) {
            for (int i = 1; i <= rows; i++) {
                for (char c = 'A'; c <= 'Z'; c++) {
                    string cell = string(1, c) + to_string(i);
                    cellMap[cell] = 0;
                }
            }
        }
    }

    void setCell(string cell, int value) {
        cellMap[cell] = value;
    }

    void resetCell(string cell) {
        cellMap[cell] = 0;
    }

    int getValue(string formula) {
        vector<string> tokens = extractTokens(formula);
        auto resolve = [&](string token) {
            if (all_of(token.begin(), token.end(), ::isdigit)) {
                return stoi(token);
            }
            return cellMap[token];
        };
        return resolve(tokens[0]) + resolve(tokens[1]);
    }
};

File: Sep/Day19/design-spreadsheet.py
import re
class Spreadsheet:
    # TC: 0(1) , SC: O(m*26)
    def __init__(self, rows: int):
        self.cellMap = {}
        rowsArray = [str(i) for i in range(1, rows+1)]
        for i in rowsArray:
            for j in 'ABCDEFGHIJKLMNOPQRSTUVWXYZ':
                self.cellMap[j+i] = 0
        
    def extractTokens(self, formula: str):
        '''
        [A-Z]+\d+ → matches cell reference (column letters + row number).
        \d+ → matches numbers.
        '''
        # Match either a cell reference (letters+digits) OR just numbers
        return re.findall(r'[A-Z]+\d+|\d+', formula)

    def setCell(self, cell: str, value: int) -> None:
        self.cellMap[cell] = value
        

    def resetCell(self, cell: str) -> None:
        self.cellMap[cell] = 0
        

    def getValue(self, formula: str) -> int:
        formulaArr = self.extractTokens(formula)
        firstNum = self.cellMap[formulaArr[0]] if formulaArr[0] in self.cellMap else int(formulaArr[0])
        secondNum = self.cellMap[formulaArr[1]] if formulaArr[1] in self.cellMap else int(formulaArr[1])

        return firstNum+secondNum



class Spreadsheet:

    def __init__(self, rows: int):
        self.cellMap = defaultdict(int)

    def extractTokens(self, formula: str):
        '''
        [A-Z]+\d+ → matches cell reference (column letters + row number).
        \d+ → matches numbers.
        '''
        return re.findall(r'[A-Z]+\d+|\d+', formula)

    def setCell(self, cell: str, value: int) -> None:
        self.cellMap[cell] = value

    def resetCell(self, cell: str) -> None:
        if self.cellMap[cell]:
            del self.cellMap[cell]

    def getValue(self, formula: str) -> int:
        formulaArr = self.extractTokens(formula)
        def resolve(token: str) -> int:
            if token.isdigit():
                return int(token)
            return self.cellMap[token]
        firstNum = resolve(formulaArr[0])
        secondNum = resolve(formulaArr[1])
        return firstNum + secondNum



# Your Spreadsheet object will be instantiated and called as such:
# obj = Spreadsheet(rows)
# obj.setCell(cell,value)
# obj.resetCell(cell)
# param_3 = obj.getValue(formula)